CHT Application Testing Tutorial (#698)
medic/cht-docs: content/en/apps/tutorials/application-tests.md

```js
describe('PNC form tests', () => {
  before(async () => { return await harness.start(); });
  after(async () => { return await harness.stop(); });
```

Any logic that you want to execute before and after each test can be placed in the `beforeEach()` and `afterEach()` hooks. Clear the state of the harness and set the date to a fixed value before each test run:
```js
  beforeEach(
    async () => {
      await harness.clear();
      await harness.setNow('2000-01-01');
    });
```

After each test run, assert that there are no errors in the console.
```js
  afterEach(() => {
    expect(harness.consoleErrors).to.be.empty;
  });
```

If you want to learn more about these hooks, refer to this [Mocha resource](https://mochajs.org/#hooks). Feel free to customize the hooks as you see fit.

Let's look at a more detailed example. <a name="assessment-form-test">Here</a> is a test case for the Assessment form that was covered in the [previous tutorial]({{< ref "apps/tutorials/app-forms" >}}):

{{< highlight js "linenos=table,hl_lines=26 31 36,anchorlinenos=true,lineanchors=assessment-form-test" >}}
  it('unit test confirming assessment with cough since 7 days', async () => {

>     // Load the assessment form and fill 'yes' on the first page and '7' on the second page

Reviewer: Can you please split this comment into two lines? Something like below. I just noticed that it exceeds the code block box unlike other code snippets.

    // Load the assessment form and fill 'yes' on the first page 
    // and '7' on the second page

Author: Done, thanks. :)

1. From the previous tutorials, you should have a [functioning CHT instance with `cht-conf` installed locally]({{< ref "apps/tutorials/local-setup" >}}) and a [project folder set up]({{< ref "apps/tutorials/local-setup#3-create-and-upload-a-blank-project" >}}).
2. If your `package.json` file does not already have them, add `cht-conf-test-harness`, `chai`, and `mocha` by running this in your command-line:
    ```shell
    npm install --save cht-conf-test-harness chai mocha
    ```
3. Also, add the following scripts to `package.json`, if not already present:

    ```json
    "scripts": {
        "test": "npm run compile-forms && npm run compile-app-settings && npm run unittest",
        "compile-app-settings": "npx cht --no-check compile-app-settings",
        "compile-forms": "npx cht --no-check convert-app-forms convert-contact-forms",
        "unittest": "mocha test/**/*.spec.js --timeout 20000"
      }
    ```
    After adding these scripts, you are able to run the tests by running one of these commands from the command-line:

>     <table>

Reviewer: I think table can be replaced with below structure (Reference: https://docs.communityhealthtoolkit.org/contribute/docs/style-guide/#use-markdown-notation)
|Command|Description|
|---|---|
|npm run test|Compiles the app settings, converts the forms, then runs the tests. <sup>[recommended]|
|npm run unittest|Only runs the tests.|

Author: I used HTML table because the `npm run unittest` was wrapping. But that might not be an enough reason. I've changed it to markdown. Thanks.

Author: Also, if you know a way to not wrap it, please let me know. It is looking like this:
<img width="249" alt="image" src="https://user-images.githubusercontent.com/20041732/183677982-500bafed-eb53-4411-80d6-77e6be393309.png">

      </tr>
    </table>

4. Create a folder in the project root where you keep the tests. You can name the folder yourself. For convenience, we will name it as `test`.
5. Create a file `harness.defaults.json` at the root of your project. This is the default configuration file for the harness. Here you can specify the default user, preloaded contacts and reports, and add other settings. Here's an [example](https://github.com/medic/cht-conf-test-harness/blob/master/harness.defaults.json.example) file to get you started. You can read more about it [here](https://docs.communityhealthtoolkit.org/cht-conf-test-harness/global.html#HarnessInputs).

## Writing a test for CHT App

We start by adding a file where a group of related tests are written. It is customary to keep one test file per individual unit of the component. For example, in the default config, all tests for the pregnancy form are in the [pregnancy.spec.js](https://github.com/medic/cht-core/blob/master/config/default/test/forms/pregnancy.spec.js) file.

Let's introduce some important sections from a typical test file:

```js
const TestRunner = require('cht-conf-test-harness');
const harness = new TestRunner();
```

> We get an instance of `harness` once and use it throughout the test file.

Reviewer: suggesting replacing we with you.
```suggestion
You get an instance of `harness` and the same instance is used throughout the test file.
```

> We can also pass certain options to the `TestRunner()` when instantiating:

Reviewer: suggesting replacing we with you and options with parameters.
```suggestion
You can also pass parameters to the `TestRunner()` when instantiating:
```

Depending on the form design, the number of inputs to be filled can be large. The inputs are often repeated within a single test or across multiple tests with little or no variation. It is a good idea to keep them in a separate file and refer them from the tests as required. You can also introduce some variations in the inputs using function parameters.

Example: `form-inputs.js`
```js {linenos=table}
module.exports = {
  assessments: {
    no_cough: [
      ['no']
    ],
    cough: (days) => [
      ['yes'],
      [ days ]
    ]
  }
};
```

In the test below, the inputs are used from the `form-inputs.js` file above:
```js
const { assessments } = require('../form-inputs');
...
const result = await harness.fillForm('assessment', ...assessments.cough('3'));
```

The test files are usually grouped in folders to read and run them easily. One way of grouping them is by creating folders for each of the components that can be tested: forms, contact summary, tasks, and targets.

> ![alt_text](tests-dir.png "Test directory structure")

Reviewer: Please replace with {{< figure src="tests-dir.png" link="tests-dir.png" alt="" title="" >}}
Style guide recommends above format instead of ![Alt Text]

Author: I agree. Thanks!

### Testing Forms

> <table>

Reviewer: Similar to above comments on table,  please refer to: https://docs.communityhealthtoolkit.org/contribute/docs/style-guide/#use-markdown-notation

Author: I had used HTML table here to use colspan, but I think we don't need to use it. Converted to markdown. Thanks!

  <tr>
    <th colspan='2'> What do you test? </th>
  </tr>
  <tr>
   <td>Minimum:</td>
   <td>Filling a form with the most common options should not result in any error.</td>
  </tr>
  <tr>
   <td>Ideal:</td>
   <td>All input combinations and constraints are tested with the report fields.</td>
  </tr>
</table>

[Previous example](#assessment-form-test) demonstrates test for the app form. More test cases can be added by changing the inputs.

You can also test contact forms using test harness. To fill a contact form, use <code>[fillContactForm(contactType, …answers)](https://docs.communityhealthtoolkit.org/cht-conf-test-harness/Harness.html#fillContactForm)</code>.

Example:

> ```js {linenos=table}

Reviewer: Do we need {linenos=table}? I am not sure but I noticed some code snippet has line numbers and some don't. Just a suggestion, may we can exclude from all others where we are not referencing line numbers of the code snippet just to make it consistent?

Author: Excluded 1`linenos` from everywhere where not referencing. Thanks!

const result = await harness.fillContactForm(
  'district_hospital',
  ['new_person', 'William Mars', '1990-08-06', '+1234567891', 'female'],
  ['yes']
);
expect(result.errors).to.be.empty;
expect(result.contacts).excluding(['_id', 'meta']).to.deep.eq([
  {
    ...
  }
]);
```

When filling a form, you can also test the field constraints. The example below asserts that a form does not accept birth date in the future:
```js {linenos=table}
it(`Throws validation error when birth date is in future`, async () => {
  const result = await harness.fillForm('delivery', ['yes', '2090-01-02']);
  expect(result.errors.length).to.equal(1);
  expect(result.errors[0]).to.include({
    msg: 'Date cannot be in the future!'
  });
});
```

> > Note: If a form triggers a task, some use cases of the form can be tested when testing the task later.

Reviewer: Please replace with {{% alert title="Note" %}} If a form triggers a task, some use cases of the form can be tested when testing the task later.{{% /alert %}}
Reference: https://docs.communityhealthtoolkit.org/contribute/docs/style-guide/#notes-and-tips

---
### Testing Contact Summary

> <table>

Reviewer: Similar to above comments on table, please refer to: https://docs.communityhealthtoolkit.org/contribute/docs/style-guide/#use-markdown-notation

Author: Converted. Thanks!

```js {linenos=table}:
// Load the assessment form and fill in 'yes' on the first page and '7' on the second page
const result = await harness.fillForm('assessment', ['yes'], ['7']);

// Verify that the form successfully got submitted
expect(result.errors).to.be.empty;

const contactSummary = await harness.getContactSummary();

// Verify that there is one card
expect(contactSummary.cards).to.have.property('length', 1);

// Verify the fields on the card
expect(contactSummary.cards[0].fields).to.deep.equal(
  [
    {
      label: 'contact.profile.most_recent_assessment.date',
      value: result.report.reported_date,
      width: 6,
      filter: 'simpleDate'
    },
    { label: 'contact.profile.cough', value: 'yes', width: 6 },
  ]
);
```
If you  follow [this code sample]({{< ref "apps/reference/contact-page#code-samples" >}}) to create the pregnancy condition card, the pregnancy context can be tested this way:
```js {linenos=table}
const summaryContext = harness.getContactSummary().context;
expect(summaryContext).to.include({
  pregnant: true
});
```
---

### Testing Tasks

> <table>

Reviewer: Similar to above comments on table, please refer to: https://docs.communityhealthtoolkit.org/contribute/docs/style-guide/#use-markdown-notation

Author: Converted. Thanks!

    {
      "_id": "chw_id",
      "name": "CHW",
      "date_of_birth": "1971-03-10",
      "phone": "+17782475555",
      "reported_date": 1550559625153,
      "type": "person",
      "parent": {
        "_id": "chw_area_id",
        "parent": {
          "_id": "district_id"
        }
      }
    },
    {      
      "_id": "chw_area_id",
      "type": "contact",
      "contact_type": "chw_area"
    },
    {
      "_id": "patient_id",
      "type": "contact",
      "contact_type": "patient",
      "name": "Patient Name",
      "role": "patient",
      "date_of_birth": "1970-07-09",
      "sex": "female",
      "reported_date": 1652868491000,
      "patient_id": "patient_id",
      "parent": {
        "_id": "patient_parent_id",
        "parent": {
          "_id": "chw_area_id",
          "parent": {
            "_id": "district_id"
          }
        }
      }
    }
  ]
}
{{< /highlight >}}

Please note that the `reported_date` above stores the epoch timestamp in milliseconds when the document was first created. You can use this external website to convert the timestamp to and from a human readable date: [https://www.epochconverter.com](https://www.epochconverter.com/).

> After setting the harness defaults, we can now test the task:

Reviewer: suggesting replacing we with you.
```suggestion
After setting the harness defaults, you can now test the task:
```

```js {linenos=true}
// Get the task by title
const tasks = await harness.getTasks({title: 'First Assessment'});

//Make sure that we have the task
expect(tasks).to.have.property('length', 1);

// Load the task action and fill it.
const result = await harness.loadAction(tasks[0], ['yes'], ['7']);

// Verify that the form successfully got submitted
expect(result.errors).to.be.empty;

// Verify that the task is no longer visible
expect(await harness.getTasks()).to.be.empty;
```

> If a task is triggered by a report instead, then we can test it by first filling the app form to create the report:

Reviewer: suggesting removal of we.
```suggestion
If a task is triggered by a report, then fill in the app form to create report before test:
```

```js {linenos=table}
it('followup schedule', async () => {
  const result = await harness.fillForm(...pncFollowups.registerBirth('2000-01-01', '2000-02-01'));
  expect(result.errors).to.be.empty;

  /* Task should appear 7 days after birth for follow up */
  await harness.flush(7);
  const tasks = await harness.getTasks();
  expect(tasks).to.have.property('length', 1);
   
  /* Complete the task, task should disappear */
  await harness.loadAction(tasks[0].emission.actions[0]);
  const followupResult = await harness.fillForm(...pncFollowups.followup.healthy);
  expect(followupResult.errors).to.be.empty;
  expect(await harness.getTasks()).to.be.empty;

  /* Confirm a task appears again on the schedule */
  await harness.flush(21);
  expect(await harness.getTasks()).to.have.property('length', 1);
});
```

> Sometimes, we [pass other data](https://docs.communityhealthtoolkit.org/apps/guides/tasks/pass-data-to-form/) from a task to the action form using [modifyContent ](https://docs.communityhealthtoolkit.org/apps/guides/tasks/pass-data-to-form/#modifycontent)attribute of a task. We can also verify that these data are present in the task.

Reviewer: suggesting replacing we with you.
```suggestion
You may [pass other data](https://docs.communityhealthtoolkit.org/apps/guides/tasks/pass-data-to-form/) from a task to the action form using [modifyContent ](https://docs.communityhealthtoolkit.org/apps/guides/tasks/pass-data-to-form/#modifycontent)attribute of a task. You can also verify that these data are present in the task.
```

```js
    expect(tasks[0].emission.actions[0].content).to.include({
      t_follow_up_count: '5',
      t_delivery_date:'2000-01-01',
    });

```
--- 
### Testing Targets

Testing a target is relatively straightforward. Add a report or contact that increments a target, then check the target values.

> <table>

Reviewer: Similar to above comments on table, please refer to: https://docs.communityhealthtoolkit.org/contribute/docs/style-guide/#use-markdown-notation

Author: Converted. Thanks!

> Testing a target is relatively straightforward. We add a report or contact that increments a target, then we check the target values.

Reviewer: suggesting removal of we.
```suggestion
Testing a target is relatively straightforward. Add a report or contact that increments a target, then check the target values.
```

  <tr>
    <th colspan='2'> What do you test? </th>
  </tr>
  <tr>
    <td>Minimum</td>
    <td>
      Trigger incrementing of the target <br />
      Ensure target doesn’t increment when it shouldn’t
    </td>
  </tr>
  <tr>
    <td>Ideal</td>
    <td>
      One test for each user scenario<br />
      Ensure proper deduplication (particularly for those with emitCustom)
    </td>
  </tr>
</table>

We use [`harness.getTargets`](https://docs.communityhealthtoolkit.org/cht-conf-test-harness/Harness.html#getTargets) to check the state of targets. It returns a [`Target`](https://docs.communityhealthtoolkit.org/cht-conf-test-harness/global.html#Target) object which corresponds to the [targets schema](https://docs.communityhealthtoolkit.org/core/overview/db-schema/#targets).

> To test the first two targets created in the [targets tutorial]({{< ref "apps/tutorials/targets" >}}), we can use this code:

Reviewer: suggesting removal of we.
```suggestion
To test the first two targets created in the [targets tutorial]({{< ref "apps/tutorials/targets" >}}), use this code:
```

    const thisMonth = await harness.getTargets({ type: 'assessments-this-month' });
    expect(thisMonth).to.have.property('length', 1);

    //The number of assessments in this month should be 1
    expect(thisMonth[0]).to.nested.include({ 'value.pass': 1, 'value.total': 1 });

    let allTime = await harness.getTargets({ type: 'assessments-all-time' });
    expect(thisMonth).to.have.property('length', 1);

    //The number of all time assessments should be 1
    expect(thisMonth[0]).to.nested.include({ 'value.pass': 1, 'value.total': 1 });

    //Go to next month by adding 2 days (Jan 30 to Feb 1)
    harness.flush(2);

    const nextMonth = await harness.getTargets({ type: 'assessments-this-month' });
    expect(nextMonth).to.have.property('length', 1);

    //The number of assessments this month now should be reset to 0
    expect(nextMonth[0]).to.nested.include({ 'value.pass': 0, 'value.total': 0 });

    allTime = await harness.getTargets({ type: 'assessments-all-time' });
    expect(allTime).to.have.property('length', 1);

    //The number of all time assessments should still be 1
    expect(allTime[0]).to.nested.include({ 'value.pass': 1, 'value.total': 1 });
  });
```

> For targets with `type: 'percent'`, we might want to check for more properties:

Reviewer: suggesting replacing we with you.
```suggestion
For targets with `type: 'percent'`, you might want to check for more properties:
```
